Segment backed broadcast join IndexedTable

clintropolis opened the pull request:
### Description
Related to #9953, this PR adds a simple naive implementation of an `IndexedTable` backed by Druid segments ingested with some extra information to identify which columns are 'join keys' at ingestion time, through the use of a custom segmentizer factory, via the changes done in #9957:
```
      "indexSpec": {
        "segmentLoader": {
          "type": "broadcastJoinableMMapSegmentFactory",
          "keyColumns": ["col1", "col2", "col3"]
        }
      }
```
This is probably longer term done better as being part of `ColumnCapabilities`, maybe a `isJoinKey` method, but since there might be other optimizations/customizations to the segment we are writing out to use for joins, I wanted to defer this change to future work to determine if there is a better segment format for this use case.

The implementation is based on and similar to `RowBasedIndexedTable`, but is fed with a `QueryableIndexSegment` which is scanned and the key columns are read to create the value indexes of the key columns, which are stored in heap memory in a list of hashmaps. The intention is that these will only be created for datasources that are configured to have a 'broadcast' load rule, which will propagate the entire dataset across the cluster, allowing the table to be used for more efficient joins.

The `SegmentManager` has been modified to extend `DataSourceState` to also contain a map of `SegmentId` to another newly added class in this PR, `ReferenceCountingIndexedTable` (related #9982), which ties the lifecycle of these tables to the segment that is backing them. 

A new `BroadcastTableJoinableFactory` has been added which is injected with the `SegmentManager` to allow fetching the indexed tables from these broadcast segments to construct into a `Joinable`, binding to `GlobalTableDatasource` introduced in #10020.

Since it is an extension point, I have also reworked `JoinableFactory` a bit to allow binding multiple implementations to the same `DataSource` class, and so now is handled with 2 bindings, a `MapBinder<Class<? extends JoinableFactory>, Class<? extends DataSource>>` and a `Multibinder<JoinableFactory>` to provide `MapJoinableFactory` so it can build a `SetMultimap<Class<? extends DataSource>, JoinableFactory>`, which it will now try each factory when building a `Joinable`, returning the first that produces results. 

Unit tests and an integration test have been added to test this functionality, and the integration test also should cover much of the broadcast changes done in service of #9953.

This PR adds no new documentation yet, since **the functionality should still be considered experimental**, but I'll consider how to best document it before the next release and do it in a future follow-up PR. 

**Also note that there are currently some limitations with `BroadcastTableJoinableFactory`, most obviously that it only supports joins to single segment datasources.** The `SegmentManager` itself does not care and will load any number of segments and tables, but joinable construction will fail. Part of the reason for this limitation is that I'm not certain the best way to resolve this, we could either make a composite IndexedTable that can in essence 'page' through it's sub-tables, or, we make a version of the join matcher that can handle a collection of tables, (or something else entirely?). It needs further investigation.

<hr>

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] added integration tests.
- [x] been tested in a test Druid cluster.

<hr>

##### Key changed/added classes in this PR
 * `BroadcastSegmentIndexedTable`
 * `ReferenceCountingIndexedTable`
 * `BroadcastTableJoinableFactory`
 * `BroadcastJoinableMMappedQueryableSegmentizerFactory`
 * `SegmentManager`
 * `MapJoinableFactory`

jihoonson commented on processing/src/test/java/org/apache/druid/segment/join/table/BroadcastSegmentIndexedTableTest.java:
```diff
@@ -0,0 +1,274 @@
+  @Before
+  public void setup() throws IOException, SegmentLoadingException
+  {
+    NullHandling.initializeForTests();
```
Seems unnecessary since this test extends `InitializeNullHandlingTest`.

clintropolis commented:
I made some fixes to this PR because while I was testing some stuff I ran into a leak of buffers, because the columns were not being closed that were created by creating the `Reader`. In response to this, I've modified `HashJoinSegmentStorageAdapter` to feed a `Closer` that runs in the sequence baggage similar to `QueryableIndexCursorSequenceBuilder`, and thread it into the join matcher/cursors down into `IndexedTableColumnSelectorFactory` and into `IndexedTableColumnValueSelector` and `IndexedTableDimensionSelector`, so that the `IndexedTable.Reader` they create can be closed when the cursors sequence is processed.

I've also added a new method to the `IndexedTable` interface:
```java
  @Nullable
  default ColumnSelectorFactory makeColumnSelectorFactory(ReadableOffset offset, boolean descending, Closer closer)
  {
    return null;
  }
```

to allow it to directly provide `ColumnSelectorFactory` to `IndexedTableJoinMatcher` instead of using `IndexedTableColumnSelectorFactory`. This allows `BroadcastSegmentIndexedTable` to supply a `QueryableIndexColumnSelectorFactory` directly from the underlying segment instead of all the indirection of using a `Reader` through the standard selectors. The performance improvement seems pretty significant (from some benchmarks I've been playing with that are not in this PR, will add later, projection '6' in this case is selecting 3 columns from the rhs table):

before:
```
Benchmark                                                 (lhsJoinColumn)  (projection)  (rhsJoinColumn)  (rowsPerSegment)  (rowsPerTableSegment)  (tableType)  Mode  Cnt    Score    Error  Units
IndexedTableBenchmark.hashJoinCursorColumnValueSelectors          dimZipf             6       dimUniform             50000                  50000      segment  avgt    5   77.403 ± 57.426  ms/op
```

after:
```
Benchmark                                                 (lhsJoinColumn)  (projection)  (rhsJoinColumn)  (rowsPerSegment)  (rowsPerTableSegment)  (tableType)  Mode  Cnt    Score    Error  Units
IndexedTableBenchmark.hashJoinCursorColumnValueSelectors          dimZipf             6       dimUniform             50000                  50000      segment  avgt    5  20.060 ± 1.627  ms/op
```

I'm sure there are a lot more improvements that could be made, but will save that for future work.

jon-wei commented on processing/src/main/java/org/apache/druid/segment/join/MapJoinableFactory.java:
```diff
@@ -19,49 +19,64 @@
+      Optional<Joinable> candidate = factory.build(dataSource, condition);
+      if (candidate.isPresent()) {
+        if (maybeJoinable.isPresent()) {
+          throw new ISE("Multiple joinable factories are valid for table[%s]", dataSource);
```
Can you explain why this is an error condition? From the comment at https://github.com/apache/druid/pull/10224/files#diff-51b09b44a8330d19728379390eef60a0R35 it sounds like multiple valid should be okay

clintropolis replied:
That comment isn't super clear, will add some more details. Multiple factories can be bound to the same class of datasource (because factory is an extension point), but not to the same exact datasource (because only 1 factory can create the joinable to use in the query). Having a mix of joinables in the same datasource is an invalid state, so this explodes to prevent undefined behavior instead of just using whichever is first in the set.

clintropolis commented on processing/src/main/java/org/apache/druid/segment/join/table/RowBasedIndexedTable.java:
```diff
@@ -69,10 +68,6 @@ public RowBasedIndexedTable(
     this.keyColumns = keyColumns;
     this.version = version;
 
-    if (new HashSet<>(keyColumns).size() != keyColumns.size()) {
```
Ah, no, #9287 refactored the signature from using a `List<String>` to a `Set<String>`, which I think makes this check pointless.

jihoonson commented on processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinable.java:
```diff
@@ -97,41 +103,49 @@ public JoinMatcher makeJoinMatcher(
+      }
+    }
+    finally {
+      CloseQuietly.close(closer);
```
Can we use try-with-resources instead of `CloseQuiently`? It would be more aligned with https://github.com/apache/druid/pull/10247 as well.

clintropolis replied:
removed all usage of `CloseQuietly` in this PR

jihoonson commented on server/src/main/java/org/apache/druid/segment/join/BroadcastTableJoinableFactory.java:
```diff
@@ -0,0 +1,79 @@
+      return segmentManager.getIndexedTables(analysis).map(tables -> {
+        Iterator<ReferenceCountingIndexedTable> tableIterator = tables.iterator();
+        if (!tableIterator.hasNext()) {
+          return null;
```
Should it return `Optional.empty()` instead?

clintropolis replied:
No, since this is inside a `Optional.map`, the function needs to return the thing that goes in the `Optional`

jihoonson replied:
Oops, you are correct.

jihoonson commented on server/src/main/java/org/apache/druid/server/SegmentManager.java:
```diff
@@ -258,6 +316,11 @@ public void dropSegment(final DataSegment segment)
+
+              final ReferenceCountingIndexedTable oldTable = dataSourceState.tablesLookup.remove(segment.getId());
+              if (oldTable != null) {
+                oldTable.close();
```
Hmm, should we use `Closer` to close `oldTable` and `oldQueryable` so that both can be closed even when any potential errors is thrown?

clintropolis replied:
changed to use `Closer`

clintropolis commented on processing/src/main/java/org/apache/druid/segment/join/HashJoinEngine.java:
```diff
@@ -51,14 +52,16 @@ private HashJoinEngine()
    * column with this name.
    */
-  public static Cursor makeJoinCursor(final Cursor leftCursor, final JoinableClause joinableClause)
+  public static Cursor makeJoinCursor(final Cursor leftCursor, final JoinableClause joinableClause, boolean descending, final Closer closer)
```
changed, also fixed formatting because line was too long

clintropolis commented:
Thanks for the review @jon-wei and @jihoonson :metal:

abhishekagarwal87 commented on processing/src/main/java/org/apache/druid/segment/join/table/BroadcastSegmentIndexedTable.java:
```diff
@@ -0,0 +1,255 @@
+  @Override
+  public Reader columnReader(int column)
+  {
+    if (column < 0 || rowSignature.getColumnName(0) == null) {
```
```suggestion
    if (column < 0 || rowSignature.getColumnName(column) == null) {
```

clintropolis replied:
oops, good catch, though it accidentally will throw the same index out of bounds exception a few lines later

```java
    final BaseColumn baseColumn = queryableIndex.getColumnHolder(rowSignature.getColumnName(column)).getColumn();
```

However, this check isn't really a good check since it is inconsistent which exception will be thrown, so I've added a `contains` method to `RowSignature` that works with column position to ensure that the index is in range so the exception can be consistently be the IAE being thrown here.
